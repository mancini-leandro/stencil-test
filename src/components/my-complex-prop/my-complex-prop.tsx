import { Component, Host, h, Prop, Watch } from '@stencil/core';

@Component({
  tag: 'my-complex-prop',
  styleUrl: 'my-complex-prop.css',
  shadow: true,
})
export class MyComplexProp {
  /**
   * Values prop
   */
  @Prop({ mutable: true })
  values: Array<string> = [];

  items: Array<string> = [];

  toUpper(items: Array<string>) {
    return items && items.length > 0 ? items.map(item => item.toUpperCase()) : [];
  }

  @Watch('values')
  onValuesChange(newValue: any) {
    this.items = this.toUpper(newValue);
  }

  componentWillLoad() {
    this.items = this.toUpper(this.values);
  }

  render() {
    return (
      <Host>
        <div class="values">
          {this.items.map((item) => {
            return <div class="values__item">
              <span>{item}</span>
            </div>
          })}
        </div>
      </Host>
    );
  }

}
