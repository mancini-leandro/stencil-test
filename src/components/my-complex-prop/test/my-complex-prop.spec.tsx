import { newSpecPage } from '@stencil/core/testing';
import { MyComplexProp } from '../my-complex-prop';

describe('my-complex-prop', () => {
  it('should change to upper case', () => {
    const comp = new MyComplexProp();
    const res = comp.toUpper(['aaa', 'bbb', 'ccc']);

    expect(res).toEqual(['AAA', 'BBB', 'CCC']);
  });

  it('should render', async () => {
    const page = await newSpecPage({
      components: [MyComplexProp],
      html: `<my-complex-prop></my-complex-prop>`
    });

    page.rootInstance.values = ['aaa', 'bbb', 'ccc'];

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.items).toEqual(['AAA', 'BBB', 'CCC']);
  });

  it('should render with data', async () => {
    const page = await newSpecPage({
      components: [MyComplexProp],
      html: '<div></div>'
    });

    const comp = page.doc.createElement('my-complex-prop');

    (comp as any).values = ['Leandro', 'Mancini'];

    page.root.appendChild(comp);

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.items).toEqual(['LEANDRO', 'MANCINI']);
  });
});
