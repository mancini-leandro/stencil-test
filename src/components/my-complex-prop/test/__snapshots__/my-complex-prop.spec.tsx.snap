// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`my-complex-prop should render 1`] = `
<my-complex-prop>
  <mock:shadow-root>
    <div class="values">
      <div class="values__item">
        <span>
          AAA
        </span>
      </div>
      <div class="values__item">
        <span>
          BBB
        </span>
      </div>
      <div class="values__item">
        <span>
          CCC
        </span>
      </div>
    </div>
  </mock:shadow-root>
</my-complex-prop>
`;

exports[`my-complex-prop should render with data 1`] = `
<my-complex-prop>
  <mock:shadow-root>
    <div class="values">
      <div class="values__item">
        <span>
          LEANDRO
        </span>
      </div>
      <div class="values__item">
        <span>
          MANCINI
        </span>
      </div>
    </div>
  </mock:shadow-root>
</my-complex-prop>
`;
